tools/roxvet/common/traverse.go

```diff
@@ -0,0 +1,62 @@
+package common
+
+import (
+	"go/ast"
+
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+// FilteredPreorder calls `inspector.Preorder(nodeTypes, fn)`, but filters out all files that do not pass the given
+// fileFilter.
+func FilteredPreorder(inspector *inspector.Inspector, fileFilter FileFilter, nodeTypes []ast.Node, fn func(n ast.Node)) {
+	effTypes := nodeTypes[:len(nodeTypes):len(nodeTypes)]
+	hadFile := hasFile(nodeTypes)
+	if !hadFile {
+		effTypes = append(effTypes, (*ast.File)(nil))
+	}
+	skip := false
+	inspector.Preorder(effTypes, func(n ast.Node) {
+		if f, ok := n.(*ast.File); ok {
+			skip = !fileFilter(f)
```

Comment:
Just for my own understanding, we come across some file which we do not care about. `skip` will tell us to ignore this file plus any other child file (I'm assuming that just means skipping a directory, sub directories, and any files in that tree). `skip` will be reset to `false` once we exit that tree and look at a tree we do care about. Is this right?

Reply:
Not quite, in the AST representation, there are no "child files". All the children of the `*ast.File` are the AST elements contained in the `.go` file, such as imports, `var (...)` blocks, function declarations etc. In the preorder traversal, first the `*ast.File` node will be emitted, followed by all its children. Therefore, everything that comes after an `*ast.File` that is not itself an `*ast.File` belongs to that file, another `*ast.File` node signals the beginning of a new file.